site/content/blog/2020-03-15-instrumenting-go-code-via-ast-2.md

```diff
@@ -0,0 +1,186 @@
+title: "Instrumenting Go code via AST, Part 2"
+slug: instrumenting-go-code-via-ast-2
+series: "AST"
+date: 2020-03-15T00:00:00-04:00
+author: Eli Yukelzon
+github: reflog
+community: eli.yukelzon
+---
+
+## Welcome!
+
+This is the second part of our AST blog post series, expanding on the subject of utilizing Go AST libraries to automate and improve your workflow.
+
+In this post I'll discuss a rather common problem that comes up while working with Go code and the way we've solved it by sprinkling a little bit of AST magic dust. Let's dive in.
+
+## Problem: A `struct` with no `interface`
+
+Let's say you are working on a large code base that was not built with `interfaces` in mind, meaning, there are `structs` and methods attached to those `structs`, but there is no `interface` describing it. This is a perfectly valid approach when you don't need to mock/stub the method implementations provided by that `struct`, or when there isn't more than one implementation of the same 'contract'. When these things **are** needed - we need to provide an `interface`.
+
+Here's a small code snippet to demonstrate:
+
+```go
+
+type Person struct {
+	Name string
+}
+
+func (p Person) Hello() string {
+	return "Hello: " + p.Name
+}
+
+type Animal struct {
+	Legs int
+}
+
+func (a Animal) Hello() string {
+	return fmt.Sprintf("I have %d legs!", a.Legs)
+}
+
+func main() {
+	p := Person{Name: "Fred"}
+	a := Animal{Legs: 4}
+	// ...	
+}
+```
+
+In this example both `Person` and `Animal` have the same `Hello()` method. If we wanted to store a list of **both** `Person` and `Animal` `structs`, we would have to define it as:
+```go
+list := []interface{}{p,a}
+```
+But this way we lose the type information of the list elements.
+This is where `interfaces` come in. Since both `Person` and `Animal` _implement_ a method with the same signature, we can extract that signature into an `interface` and use it for storing items in a list:
+```go
+type interface Hello {
+	Hello() string
+}
+// ...
+list := []Hello{p,a}
+fmt.Printf("Person: [%v] Animal: [%v]\n", list[0].Hello(), list[1].Hello())
+```
+
+Awesome. Now let's say the `struct` you are extracting the `interface` from is a big one. A really big one. With lots and lots of methods spread out in different `.go` files. Creating such an `interface` manually would be very laborious.
+
+This problem is itching to get an AST treatment. Let's get to it!
+
+## AST to the rescue!
+
+Let's break down the task at hand into smaller, digestable parts:
+
+1. Scan the source code for all methods implemented on a specific `struct`
+1. Collect all those methods (and their comments, for clarity)
+1. Create a new file that will contain an `interface` with all the collected methods inside
+1. ...
+1. Profit?
+
+> Note: usually the `interface` doesn't contain **all** of the `struct` methods, but we'll leave the topics of abstraction and better code organization for another blog post
+
+
+### Scanning the source code for all `struct`'s methods
```

Comment:
```suggestion
### Scanning the source code for all `struct` methods
```

Reply:
Not sure whether this suggestion works - are we referring to the plural or a specific struct's method?